### PR TITLE
feat(webui): add Tool Activity Panel to right sidebar

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

- Adds a **Tool Activity** panel (Wrench icon) to the webui right sidebar showing which gptme tools were used in the current session — name, call count, last args/content preview, and timestamp
- Parses assistant messages for markdown code blocks with known gptme tool names (bash, python, save, append, patch, browser, tmux, etc.) and maintains a per-session inventory sorted by call count
- Each tool row is collapsible to show the last call's args and content digest

## Motivation

Tool calls appear inline in the chat stream but vanish into scroll history. When a session runs many tool calls, it's hard to answer "what has the agent actually done?" without scrolling back. This panel surfaces that information persistently, reducing the cognitive load of understanding agent state — a key usability gap highlighted in the [HN GUI for AI agents](https://news.ycombinator.com/item?id=49119274) discussion.

## Implementation

- `webui/src/utils/toolActivity.ts`: `buildToolActivity(messages)` parses code fences from assistant messages, filtering to known gptme tool names, and returns `ToolActivityEntry[]` sorted by call count
- `webui/src/components/ToolActivityPanel.tsx`: collapsible sidebar panel with per-tool icons and content preview
- `webui/src/utils/__tests__/toolActivity.test.ts`: 11 unit tests covering parsing, deduplication, sorting, and edge cases
- `webui/src/components/rightSidebarPanels.tsx` + `webui/src/types/sidebar.ts`: register panel in existing sidebar plugin architecture

## Test plan

- [ ] Unit tests: `cd webui && npx jest src/utils/__tests__/toolActivity.test.ts` (11 tests, all passing)
- [ ] TypeScript: `cd webui && npx tsc --noEmit` (clean)
- [ ] ESLint: `cd webui && npx eslint src/utils/toolActivity.ts src/components/ToolActivityPanel.tsx` (clean)
- [ ] Manual: open a gptme session that runs bash/python/save tool calls, click Wrench icon in right sidebar, verify tool inventory updates live

### CODE REVIEW

File: webui/src/utils/toolActivity.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  const calls: ToolCall[] = [];
+  let match: RegExpExecArray | null;
+  TOOL_CALL_RE.lastIndex = 0;
+  while ((match = TOOL_CALL_RE.exec(message.content)) !== null) {
+    const toolName = match[1].toLowerCase();
+    if (!GPTME_TOOLS.has(toolName)) continue;
+    const rawArgs = match[2]?.trim() || '';
+    const args = rawArgs ? rawArgs.split(/\s+/) : [];
+    const content = match[3] || '';
+    calls.push({ tool: toolName, args, content, timestamp: message.timestamp });
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply webui/src/utils/toolActivity.ts]: Fixed in c167e9ae0. The panel now counts only parsed calls with following persisted system result messages, so a recognized fenced example in a final response is excluded. Added regression coverage for examples and partial multi-call results.

File: webui/src/components/ToolActivityPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  );
+};
+
+export const ToolActivityPanel: FC<Props> = ({ conversationId }) => {
+  const log = use$(conversations$.get(conversationId)?.data.log);
+  const activity = useMemo(() => buildToolActivity(log ?? []), [log]);
+
+  if (!log || log.length === 0) {
+    return (
+      <div className="flex h-full flex-col items-center justify-center gap-2 p-4 text-center text-muted-foreground">
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply webui/src/components/ToolActivityPanel.tsx:110]: Fixed in c167e9ae0 by making the loaded-window boundary explicit. When older messages remain, the panel says “Loaded messages only”; its empty state says “No tool calls in loaded messages” and directs the user to load older messages. It no longer claims session-wide completeness.

File: webui/src/utils/toolActivity.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  'ipython',
+  'save',
+  'append',
+  'patch',
+  'read',
+  'computer',
+  'browser',
+  'screenshot',
+  'tmux',
+]);
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply webui/src/utils/toolActivity.ts]: Fixed in c167e9ae0. Tool Activity now reuses parseToolCalls() and its canonical allowlist from toolCallParser.ts. Added coverage for mcp and gh.

**Comment by TimeToBuildBob** [comment]
@greptile-apps review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Fixed the re-review blocker in 17464e23e. Markdown calls now require system output followed by an assistant continuation; a final fenced example followed only by `TURN_POST` hook messages is excluded. Added the exact hook-output regression test.

Local verification: focused Jest suite (14/14), TypeScript typecheck, and ESLint (0 errors; existing warnings only). Fresh CI is running on the new head.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: webui/src/utils/toolActivity.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    // Markdown tool calls have no persisted call ID. A completed tool step has
+    // system output followed by the assistant's continuation. TURN_POST hooks
+    // can also emit system messages immediately after a final response, so a
+    // system message alone is not evidence that a fenced example executed.
+    let next = index + 1;
+    while (next < messages.length && messages[next].role === 'system') {
+      next++;
+    }
+    if (next >= messages.length || messages[next].role !== 'assistant') continue;
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply webui/src/utils/toolActivity.ts]: Fixed in 5c95648d3. Persisted API messages now expose the existing `call_id` on real tool-result messages, while hook/context system messages remain untagged. The panel uses that identity to count terminal results without requiring a later assistant continuation, and keeps the conservative continuation fallback for legacy logs without IDs.

Added regressions for terminal results, interleaved hook output, legacy logs, and API serialization. Verified the focused Jest suite (17/17), TypeScript typecheck, Python API tests (31/31), and all scoped pre-commit hooks.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: webui/src/utils/toolActivity.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    // system-result + assistant-continuation fallback for those logs only.
+    let next = index + 1;
+    const resultMessages: Message[] = [];
+    while (next < messages.length && messages[next].role === 'system') {
+      resultMessages.push(messages[next]);
+      next++;
+    }
+    const hasIdentifiedResults = resultMessages.some((result) => result.call_id);
+    const resultCount = hasIdentifiedResults
+      ? resultMessages.filter((result) => result.call_id).length
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply webui/src/utils/toolActivity.ts]: Fixed in 6db622050. Identified tool results now persist their producing tool in message metadata, and Tool Activity matches results to parsed calls by that provenance instead of assigning an identified count to the first calls in the batch. This handles mixed native/markdown batches and ignores untagged hook output; older logs retain the conservative call_id/continuation fallback. Added Python provenance/round-trip coverage and the two mixed-batch frontend regressions. Verified focused Python tests (22/22), Jest (19/19), TypeScript, ESLint, Ruff, and pre-commit.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: webui/src/utils/toolActivity.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      resultMessages.push(messages[next]);
+      next++;
+    }
+    const remainingByTool = new Map<string, number>();
+    for (const result of resultMessages) {
+      const tool = result.metadata?.tool?.toLowerCase();
+      if (tool) remainingByTool.set(tool, (remainingByTool.get(tool) ?? 0) + 1);
+    }
+    const identifiedResultCount = resultMessages.filter((result) => result.call_id).length;
+    let fallbackCount = remainingByTool.size > 0 ? 0 : identifiedResultCount;
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply webui/src/utils/toolActivity.ts:44]: Fixed in 003f15c2a. Generated execution-error messages now carry the same producing-tool provenance as successful identified results, so mixed success/failure batches retain both invocations. Added backend error-result coverage and the exact frontend mixed-batch regression. Verified focused Python tests (23/23), Jest (20/20), TypeScript, ESLint, Ruff, and pre-commit.

**Comment by TimeToBuildBob** [comment]
@greptileai review
